libp2p/stream_muxer/muxer_multistream.py

```diff
@@ -0,0 +1,71 @@
+from typing import Dict, List, Mapping, Type
+
+from libp2p.network.connection.raw_connection_interface import IRawConnection
+from libp2p.network.typing import GenericProtocolHandlerFn
+from libp2p.peer.id import ID
+from libp2p.protocol_muxer.multiselect import Multiselect
+from libp2p.protocol_muxer.multiselect_client import MultiselectClient
+from libp2p.protocol_muxer.multiselect_communicator import RawConnectionCommunicator
+from libp2p.security.secure_conn_interface import ISecureConn
+from libp2p.typing import TProtocol
+
+from .abc import IMuxedConn
+
+MuxerClassType = Type[IMuxedConn]
+
+# FIXME: add negotiate timeout to `MuxerMultistream`
+DEFAULT_NEGOTIATE_TIMEOUT = 60
+
+
+class MuxerMultistream:
+    """
+    MuxerMultistream is a multistream stream muxed transport multiplexer.
+    go implementation: github.com/libp2p/go-stream-muxer-multistream/multistream.go
+    """
+
+    transports: Dict[TProtocol, MuxerClassType]
+    multiselect: Multiselect
+    multiselect_client: MultiselectClient
+    order_preference: List[TProtocol]
+
+    def __init__(
+        self, muxer_transports_by_protocol: Mapping[TProtocol, MuxerClassType]
+    ) -> None:
+        self.transports = {}
+        self.multiselect = Multiselect()
+        self.multiselect_client = MultiselectClient()
+        self.order_preference = []
+        for protocol, transport in muxer_transports_by_protocol.items():
+            self.add_transport(protocol, transport)
+
+    def add_transport(self, protocol: TProtocol, transport: MuxerClassType) -> None:
+        self.transports[protocol] = transport
+        self.multiselect.add_handler(protocol, None)
+        self.order_preference.append(protocol)
+
+    async def select_transport(self, conn: IRawConnection) -> MuxerClassType:
```

Comment:
What do you think about catching `MultiselectClientError` and `MultiselectError` here then raise and propagate `MultiselectFailure` exception?

Also applies to `SecurityMultistream.select_transport `.